windows/src/global/delphi/debug/Keyman.System.DebugLogManager.pas

```diff
@@ -0,0 +1,284 @@
+  SessionGuid: TGUID = '{FADEA67E-0EE9-452B-AF04-22E342D1227A}';
+
+function GetDebugManager(AOwner: HWND): TDebugLogManager;
+begin
+  Assert(not FTerminating);
+
+  if ShouldDebug then
+  begin
+    if not Assigned(FDebugManager) then
+      FDebugManager := TDebugLogManager.Create(AOwner);
+    Result := FDebugManager;
+  end
+  else
+    Result := nil;
+end;
+
+{ TDebugManager }
+
+const LOGSESSION_NAME: string = 'Keyman Event Trace Session';
+
+function StringBufferSize(const s: string): Integer;
+begin
+  Result := (Length(s) + 1) * sizeof(WCHAR);
+end;
+
+constructor TDebugLogManager.Create(AOwner: HWND);
+var
+  BufferSize: ULONG;
+  status: ULONG;
+begin
+  FOwner := AOwner;
+
+  FindFirstLogFileName;
+  StartNewLogFile;
+
+  // Allocate memory for the session properties. The memory must
+  // be large enough to include the log file name and session name,
+  // which get appended to the end of the session properties structure.
+
+  BufferSize := sizeof(EVENT_TRACE_PROPERTIES) + StringBufferSize(FDebugLogFileName) + StringBufferSize(LOGSESSION_NAME);
+  pSessionProperties := PEVENT_TRACE_PROPERTIES(AllocMem(BufferSize));
+
+  // Set the session properties. You only append the log file name
+  // to the properties structure; the StartTrace function appends
+  // the session name for you.
+
+  pSessionProperties.Wnode.BufferSize := BufferSize;
+  pSessionProperties.Wnode.Flags := WNODE_FLAG_TRACED_GUID;
+  pSessionProperties.Wnode.ClientContext := 1; //QPC clock resolution
+  pSessionProperties.Wnode.Guid := SessionGuid;
+  pSessionProperties.LogFileMode := EVENT_TRACE_FILE_MODE_SEQUENTIAL;
+  pSessionProperties.MaximumFileSize := 1;  // 1 MB
+  pSessionProperties.LoggerNameOffset := sizeof(EVENT_TRACE_PROPERTIES);
+  pSessionProperties.LogFileNameOffset := sizeof(EVENT_TRACE_PROPERTIES) + sizeof(LOGSESSION_NAME);
+  StrPCopy(PWideChar(PByte(pSessionProperties) + pSessionProperties.LogFileNameOffset), FDebugLogFileName);
+
+  status := StartTrace(@FSessionHandle, PWideChar(LOGSESSION_NAME), pSessionProperties^);
+  if ERROR_ALREADY_EXISTS = status then
+  begin
+    // The trace was already started, perhaps Keyman did not close down cleanly
+    // We'll stop it and restart it
+    status := ControlTraceW(FSessionHandle, PWideChar(LOGSESSION_NAME), pSessionProperties^, EVENT_TRACE_CONTROL_STOP);
+    if ERROR_SUCCESS <> status then
+      OutputDebugString(PChar('ControlTrace failed with '+IntToStr(status)));
+
+    status := StartTrace(@FSessionHandle, PWideChar(LOGSESSION_NAME), pSessionProperties^);
+  end;
+
+  if ERROR_SUCCESS <> status then
+    RaiseLastOSError(status, 'StartTrace');
+
+  // Enable the providers that you want to log events to your session.
+
+  status := EnableTraceEx(
+    @DebugLogProviderGuid,
+    @SessionGuid,
+    FSessionHandle,
+    1, // Enable
+    TRACE_LEVEL_INFORMATION,
+    0,
+    0,
+    0,
+    nil
+  );
+
+  if ERROR_SUCCESS <> status then
+    RaiseLastOSError(status, 'EnableTraceEx');
```

Comment:
It seems `EnableTraceEx` is "enabling" or "disabling", depending on the 4th parameter.

1. What about the using `'EnableTraceEx(TRUE)` to contrast with `OutputDebugString(... EnableTraceEx(FALSE)...));` in line 208?

From what I gather, the fourth parameter of `1` / `0` determines whether `EnableTraceEx` is "enabling" or "disabling".

Reply:
It's not terribly important because this error trace also returns the line number where the error occurs :)